pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	if err := c.ensureEtcdGuardDeployment(ctx, replicaCount, recorder); err != nil {
+		return err
+	}
+
+	if err := c.ensureEtcdGuardPdbDeployment(ctx, recorder); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ensureEtcdGuardDeployment if etcd quorum guard deployment doesn't exist - add it, else skip
+func (c *QuorumGuardController) ensureEtcdGuardDeployment(ctx context.Context, replicaCount int32, recorder events.Recorder) error {
+	_, err := c.kubeClient.AppsV1().Deployments(operatorclient.TargetNamespace).Get(ctx,
+		EtcdGuardDeploymentName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		recorder.Eventf("NoEtcdQuorumGuardDeployment", "%s was not found, creating one", EtcdGuardDeploymentName)
+		return c.applyDeployment(replicaCount, recorder)
+	case err != nil:
+		klog.Warningf("failed to get %s deployment err %v", EtcdGuardDeploymentName, err.Error())
+		return err
+	default:
+		return nil
+	}
+}
+
+// ensureEtcdGuardDeployment if etcd quorum guard pdb deployment doesn't exist - add it, else skip
+func (c *QuorumGuardController) ensureEtcdGuardPdbDeployment(ctx context.Context, recorder events.Recorder) error {
+	_, err := c.kubeClient.PolicyV1beta1().PodDisruptionBudgets(operatorclient.TargetNamespace).Get(ctx,
+		EtcdGuardDeploymentName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		recorder.Eventf("NoEtcdQuorumGuardPDBDeployment", "%s was not found, creating one", EtcdGuardDeploymentName)
+		return c.applyPdbDeployment(ctx, recorder)
+	case err != nil:
+		klog.Warningf("failed to get %s deployment err %v", EtcdGuardDeploymentName, err.Error())
+		return err
+	default:
+		return nil
+	}
+}
+
+// applyDeployment applies new etcd quorum guard deployment
+func (c *QuorumGuardController) applyDeployment(replicaCount int32, recorder events.Recorder) error {
+	klog.Infof("Going to create new %s deployment with replica count %d", EtcdGuardDeploymentName, replicaCount)
+
+	deployment := resourceread.ReadDeploymentV1OrDie(etcd_assets.MustAsset("etcd/quorumguard-deployment.yaml"))
+	deployment.Spec.Replicas = &replicaCount
+	newDeployment, _, err := resourceapply.ApplyDeployment(c.kubeClient.AppsV1(), recorder, deployment, -1)
+	if err != nil {
+		klog.Errorf("Failed to deploy %s, error: %v", deployment.Name, err)
+		return err
+	}
+
+	klog.Infof("New etcd quorum guard deployment is %s", newDeployment.Name)
+	return err
+}
+
+// applyPdbDeployment applies new etcd quorum guard pdb
+func (c *QuorumGuardController) applyPdbDeployment(ctx context.Context, recorder events.Recorder) error {
+	klog.Infof("Creating new PDB %q", EtcdGuardDeploymentName)
+	maxUnavailable := intstr.FromInt(pdbDeploymentMaxUnavailableDefault)
+	pdb := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      EtcdGuardDeploymentName,
+			Namespace: operatorclient.TargetNamespace,
+			Annotations: map[string]string{"include.release.openshift.io/self-managed-high-availability": "true",
```

Comment:
drop annotations they are not authoritative anymore and work to maintain accurately.